### PR TITLE
Start refactoring 3d tile and point cloud

### PR DESCRIPTION
**Pull request by gchoqueux**
## Description
* add `C3DTilesLayer` and `PointCloudLayer` and remove `preprocessDataLayer` from providers.
* move processing to the layers

## Motivation and Context
In `itowns`, 3dt tiles and point cloud  don't fit the architecture :

- a first problem, their `providers` overload the `Layers` with functions and properties. This method isn't simple and has been abandoned to perform only the role of the provider: provide data in response to a command.
- as they don't follow the architecture, it adds special treatments, especially in the `Viewer`
- they didn't the process structure `Fetch/Parse/Convert`. This structure is highly reusable and maintainable. (will be carried out in another PR)

### CODE REVIEW

**Comment by mbredif** [comment]
What's the goal/rationale for this refactoring ?

**Comment by gchoqueux** [comment]
@mbredif I've added comments in[ PR description](https://github.com/iTowns/itowns/pull/1288#issue-361541592).

File: examples/js/3dTilesHelper.js [hunk 1]
```diff
@@ -21,20 +21,24 @@ function getPickedBatchInfo(intersects) {
...
-        var interAttributes = intersects[i].object.geometry.attributes;
-        if (interAttributes && interAttributes._BATCHID) {
-            // face is a Face3 object of THREE which is a
-            // triangular face. face.a is its first vertex
-            var vertex = intersects[i].face.a;
-            // get batch id of the face
-            var batchID = interAttributes._BATCHID.array[vertex];
-            var batchTable = findBatchTable(
-                intersects[i].object);
+        var interAttributes = intersects[i].object.geometry.attributes;d
```
Comment (zarov) [inline examples/js/3dTilesHelper.js]: Remove the `d` at the end of the line

File: src/Layer/Layer.js [hunk 2]
```diff
@@ -69,6 +70,10 @@ class Layer extends THREE.EventDispatcher {
         this.defineLayerProperty('frozen', false);
 
         this.info = new InfoLayer(this);
+
+        this.whenReady = source && source.whenReady ? config.source.whenReady : Promise.resolve();
```
Comment (zarov) [inline src/Layer/Layer.js]: Use `config.source` everywhere.

File: src/Layer/C3DTilesLayer.js [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        return this.extensionsMap.has(extensionName);
+    }
+}
+
+/**
+ * Global object holding 3D Tiles extensions. Extensions must be registered
+ * with their parser by the client.
+ * @type {$3DTilesExtensions}
+ */
+export const $3dTilesExtensions = new $3DTilesExtensions();
```
Comment (zarov) [inline src/Layer/C3DTilesLayer.js:52]: If it is supposed to be a singleton, you can ditch the class to use a frozen Object like in `Cache`.

File: src/Provider/3dTilesProvider.js [hunk 4]
```diff
@@ -375,6 +125,5 @@ function executeCommand(command) {
 }
 
 export default {
-    preprocessDataLayer,
     executeCommand,
```
Comment (zarov) [inline src/Provider/3dTilesProvider.js]: Do you plan to move it in a later commit ? Asking just to know

File: src/Layer/PointCloudLayer.js [hunk 5]
```diff
@@ -0,0 +1,595 @@
...
+     * @param {number} [config.mode=MODE.COLOR] displaying mode.
+     *
+     * @param  {View}  view  The view
+     */
+    constructor(id, config, view) {
+        super(id, new THREE.Group());
+        this.isPointCloudLayer = true;
+        this.protocol = 'potreeconverter';
+
+        this.file = config.file || 'cloud.js';
```
Comment (zarov) [inline src/Layer/PointCloudLayer.js]: Not sure if a default value should be set.

File: src/Layer/PointCloudLayer.js [hunk 6]
```diff
@@ -0,0 +1,595 @@
...
+import PotreeBinParser from 'Parser/PotreeBinParser';
+import PotreeCinParser from 'Parser/PotreeCinParser';
+import PointsMaterial, { MODE } from 'Renderer/PointsMaterial';
+import Picking from 'Core/Picking';
+import Extent from 'Core/Geographic/Extent';
+import CancelledCommandException from 'Core/Scheduler/CancelledCommandException';
+
+const point = new THREE.Vector3();
+// Draw a cube with lines (12 lines).
+function cube(size) {
```
Comment (zarov) [inline src/Layer/PointCloudLayer.js:13]: This could be replaced with `THREE.BoxGeometry` ?

**Review by zarov — APPROVED** [review]
LGTM, some of my comments are to be taken into account for your later work on this.

File: examples/js/3dTilesHelper.js (hunk 1 above)
Comment (gchoqueux) [reply examples/js/3dTilesHelper.js]: ok

File: src/Layer/Layer.js (hunk 2 above)
Comment (gchoqueux) [reply src/Layer/Layer.js]: ok

File: src/Layer/PointCloudLayer.js (hunk 5 above)
Comment (gchoqueux) [reply src/Layer/PointCloudLayer.js]: i remove the default value.

File: src/Layer/PointCloudLayer.js (hunk 6 above)
Comment (gchoqueux) [reply src/Layer/PointCloudLayer.js:13]: yes, in next PR

File: src/Layer/PointCloudLayer.js (hunk 5 above)
Comment (gchoqueux) [reply src/Layer/PointCloudLayer.js]: and added mandatory config properties

File: src/Layer/PointCloudLayer.js (hunk 6 above)
Comment (gchoqueux) [reply src/Layer/PointCloudLayer.js:13]: yes, may be with OBB Helper

**Pull request merged**